### DISCUSSION TITLE
Batch process UltraStar songs to remove the vocals and create a Karaoke version

### DISCUSSION THREAD

**Discussion by Kugelblitz78** [rakuri255/UltraSinger#136] [Ideas]
Hey 
Love to use UltraSinger for existing UltraStar songs.
Would like to batch process all Songs in the Songs Folder. 
UltraSinger shall go through all the Subfolders in the Songs Folder.
Take the mp3 and split the Vocals and the Music.
Copy the TXT file and change the content. 
- add "Karaoke" at the end of the title
- change the MP3 File name to the MP3 with the Music only. 

Any Idea how to do this?

**Comment by rakuri255** [comment]
Hi. Good idea.
Simplest ways is to use python.
There would be 2 solutions.
1. A .py file that runs UltraSinger in the loop
2. Add an -i <folder> function to the UltraSinger as PR.

I prefere the 2. :) 
So everyone can use it

**Reply by Kugelblitz78** [reply]
Sounds great! 

Unfortunately I don't know python.
Is solution 2  hard to implement?

**Reply by BWagener** [reply]
I have a work in progress approach to this here

https://github.com/BWagener/UltraSinger/blob/feature/evaluation/src/UltraSingerEvaluation.py

Although it's a separate file and the original use case is different it could be adapted. I'm not working on it at the moment though.

**Reply by Kugelblitz78** [reply]
That sounds Interesting. I downloaded your repository but could not get it to run. 
Failing to import "dataclass_json"

**Reply by Kugelblitz78** [reply]
Is there any command for UltraSinger just to do the separation of the Song mp3 File into Vocals and Music (Karaoke Version) MP3s, without trying to create the TXT file with the Lyrics Beats and so on ?

**Comment by Kugelblitz78** [comment]
I found a "not so nice" but Working Multi Step Solution. 

**1st Separation)** 

As I am not a Programmer I downloaded the Program Demucs-Gui for Windows to Separate the Vocals from the Music.
https://github.com/CarlGao4/Demucs-Gui

With it I can Batch process all Folders and Subfolders and create three WAV Files.
One "Karaoke" with only the Music
one "Vocals" with only the Singing Voices
and one "music plus 25 percent vocals" so you can Practise if you are not so familiar with the Song yet.
![Demucs-gui](https://github.com/rakuri255/UltraSinger/assets/1755032/834bb71d-2364-4a47-9ee6-150edc84367f)



**2nd WAV -> MP3**
With the VLC Player I batch convert all WAV files to MP3 files and then do a search in the Explorer for WAV files in the Subfolders and delete them all.

**3rd create the TXT files**
with the help of AI (ChatGPT) the "process_karaoke.bat" Batch file was created.
that File makes a copy of the TXT file in each folder and renames it to  
SONGNAME - music plus 25 percent vocals.txt
SONGNAME - karaoke.txt
SONGNAME - vocals.txt

Works - but Would be nice to have everything in one process.

just edit the line 
`		set "rootDir=D:\Karaoke\Batch"  `
to the path where your songs are located.


and the  calls

```
		call :makeCopyAndEdit "!txtFile!" " - karaoke" " - karaoke.mp3"
		call :makeCopyAndEdit "!txtFile!" " - vocals" " - vocals.mp3"
		call :makeCopyAndEdit "!txtFile!" " - music plus 25 percent vocals" " - music plus 25 percent vocals.mp3"
```
		
how your variations are named		

**process_karaoke.bat**

```
@echo off
setlocal enabledelayedexpansion

rem Set the root directory where the script will start searching
set "rootDir=D:\songs"

rem Loop through all folders in the root directory and subdirectories
for /d /r "%rootDir%" %%A in (*) do (
    rem Process each folder
    set "folderPath=%%A"
    echo Processing folder: %%A

    rem Find and process all .txt files in the current folder
    for %%F in ("%%A\*.txt") do (
        rem Process each text file
        set "txtFile=%%F"
        echo Processing file: !txtFile!

        call :processTextFile "!txtFile!"
    )

    rem Delete the "filename" file in the current folder
    if exist "!folderPath!\filename" (
        del "!folderPath!\filename" > nul
    )
)

exit /b

:processTextFile
rem Process the specified text file
set "txtFile=%~1"
echo Editing content of file: !txtFile!

rem Create copies of the text file with specified suffixes and edit content
call :makeCopyAndEdit "!txtFile!" " - Karaoke" " - karaoke.mp3"
call :makeCopyAndEdit "!txtFile!" " - Vocals" " - vocals.mp3"
call :makeCopyAndEdit "!txtFile!" " - Music and quiet Vocals" " - music and quiet vocals.mp3"
exit /b

:makeCopyAndEdit
rem Create a copy of the text file and edit its content
set "originalFile=%~1"
set "titleSuffix=%~2"
set "mp3Suffix=%~3"
echo Creating copy with suffixes for file: !originalFile!

rem Get the base filename without extension
for %%A in ("!originalFile!") do (
    set "baseName=%%~nA"
    set "targetFile=!folderPath!\!baseName!!titleSuffix!.txt"
)

echo Making a copy of file: !originalFile!
rem Copy the original file to the target file
copy "!originalFile!" "!targetFile!" > nul
echo Copied file: !originalFile! to !targetFile!

echo Editing content of file: !targetFile!
rem Edit the content of the copied file
(for /f "usebackq tokens=*" %%A in ("!originalFile!") do (
    set "line=%%A"
    setlocal enabledelayedexpansion
    if "!line:~0,1!"=="#" (
        if "!line:~1,5!"=="TITLE" (
            echo Editing TITLE line in file: !targetFile!
            echo #TITLE:!line:~7!!titleSuffix!>>"!targetFile!.tmp"
        ) else if "!line:~1,3!"=="MP3" (
            rem Extract the MP3 filename and append the suffix
            for /f "tokens=2 delims=:" %%X in ("!line!") do (
                set "mp3File=%%X"
                echo Editing MP3 line in file: !targetFile!
                echo #MP3:!mp3File:~0,-4!!mp3Suffix!>>"!targetFile!.tmp"
            )
        ) else (
            echo !line!>>"!targetFile!.tmp"
        )
    ) else (
        echo !line!>>"!targetFile!.tmp"
    )
    endlocal
)) > nul

echo Overwriting file: !targetFile! with modified content
rem Overwrite the original file with the modified content
move /y "!targetFile!.tmp" "!targetFile!" > nul

exit /b

```



or download this file and delete the .txt at the end
[process_karaoke-works.bat.txt](https://github.com/user-attachments/files/15931237/process_karaoke-works.bat.txt)

**Reply by Kugelblitz78** [reply]
"Can you write a Python script out if it?"
[process_karaoke.py.txt](https://github.com/user-attachments/files/15931264/process_karaoke.py.txt)


Certainly! Below is a Python script that replicates the functionality of the batch script provided earlier. This Python script will recursively search through a specified directory (root_dir) and its subdirectories, locate .txt files (assumed to be karaoke lyrics files), create copies of these files with specified suffixes (- Karaoke, - Vocals only, - Karaoke and quiet Vocals), and modify the content of each copied file as required.

Please make sure to replace root_dir with the path to your karaoke files directory before running the script.

python

```
import os
import shutil

def process_text_file(txt_file, folder_path):
    title_suffixes = [" - Karaoke", " - Vocals", " - Music and quiet Vocals"]
    mp3_suffixes = [" - karaoke.mp3", " - vocals.mp3", " - music and quiet vocals.mp3"]
    
    for title_suffix, mp3_suffix in zip(title_suffixes, mp3_suffixes):
        make_copy_and_edit(txt_file, folder_path, title_suffix, mp3_suffix)

def make_copy_and_edit(original_file, folder_path, title_suffix, mp3_suffix):
    base_name = os.path.splitext(os.path.basename(original_file))[0]
    target_file = os.path.join(folder_path, f"{base_name}{title_suffix}.txt")

    # Copy the original file to the target file
    shutil.copyfile(original_file, target_file)
    
    # Read the contents of the target file
    with open(target_file, 'r', encoding='utf-8') as file:
        lines = file.readlines()
    
    # Edit the contents
    with open(f"{target_file}.tmp", 'w', encoding='utf-8') as file:
        for line in lines:
            if line.startswith("#TITLE"):
                file.write(f"#TITLE:{line[7:].strip()}{title_suffix}\n")
            elif line.startswith("#MP3"):
                mp3_filename = line[5:].strip()
                base_mp3_name = os.path.splitext(mp3_filename)[0]
                file.write(f"#MP3:{base_mp3_name}{mp3_suffix}\n")
            else:
                file.write(line)
    
    # Overwrite the original file with the modified content
    shutil.move(f"{target_file}.tmp", target_file)

def main():
    root_dir = r"D:\songs"

    for folder_path, _, files in os.walk(root_dir):
        print(f"Processing folder: {folder_path}")

        for file_name in files:
            if file_name.endswith(".txt"):
                txt_file = os.path.join(folder_path, file_name)
                print(f"Processing file: {txt_file}")
                process_text_file(txt_file, folder_path)
        
        filename_path = os.path.join(folder_path, "filename")
        if os.path.exists(filename_path):
            os.remove(filename_path)

if __name__ == "__main__":
    main()
```
This Python script performs the same tasks as the batch script: it traverses through the specified directory and its subdirectories, processes each text file, and creates copies with specified suffixes while editing the content. Let me know if you need further explanation or assistance!

Instructions for Using the Python Script:

Explanation:
process_text_file: Takes a .txt file and calls make_copy_and_edit with different suffixes.
make_copy_and_edit: Creates a copy of the file and edits the #TITLE and #MP3 lines according to the specified suffixes.
main: Walks through all subdirectories in root_dir, processes .txt files, and deletes the filename file if it exists.
How to Run:
Save the script to a file, for example, process_karaoke.py.
Make sure you have Python installed on your system.
Open a command prompt or terminal and navigate to the directory containing process_karaoke.py.
Run the script with python process_karaoke.py.
This script will process the directories and files similarly to your batch script but with the added flexibility and readability of Python.

**Comment by mARTin-B78** [comment]
I create a UltraSinger batchfile to generate Karaoke Songs out of Youtube Playlists
[batch_generate_karaoke_songs_from_youtube_playlist_with_ultrasinger.bat.txt](https://github.com/user-attachments/files/17513732/batch_generate_karaoke_songs_from_youtube_playlist_with_ultrasinger.bat.txt)


And installation Batch Scripts that install everything needed to run UltraSinger.
winget
ffmpeg
python 3.10
git
ultrasinger
musescore

JUST REMOVE THE ".TXT" at the end of the Files, Apart from the file "additional-requirements.txt" that stays as it is. [additional-requirements.txt](https://github.com/user-attachments/files/17442508/additional-requirements.txt)

Open an CMD Console in windows with admin rights.

1st-part-install-winget-ffmpeg-and-python-3.10.bat
[1st-part-install-winget-ffmpeg-and-python-3.10.bat.TXT](https://github.com/user-attachments/files/17442506/1st-part-install-winget-ffmpeg-and-python-3.10.bat.TXT)
REBOOT - cause of Python

Open an CMD Console in windows with admin rights.

2nd-part-install-git-ultrasinger-with-modules-musescore.bat
[2nd-part-install-git-ultrasinger-with-modules-musescore.bat.TXT](https://github.com/user-attachments/files/17442507/2nd-part-install-git-ultrasinger-with-modules-musescore.bat.TXT)

When Everything is installed you can use that first Batchfile batch_generate_karaoke_songs_from_youtube_playlist_with_ultrasinger.bat to Generate Karaoke Songs out of Youtube Playlists with UltraSinger

**Comment by mARTin-B78** [comment]
I update the files

I recreated the Installer for Windows made for people who are not programmers and have most tools not installed yet.

**Batch Script 1** _1st-part-install-winget-ffmpeg-and-python-3.10.bat_ download and remove the ".TXT"
[1st-part-install-winget-ffmpeg-and-python-3.10.bat.TXT](https://github.com/user-attachments/files/17591381/1st-part-install-winget-ffmpeg-and-python-3.10.bat.TXT)
1.	Check and Install **Winget**
Verifies if Winget is installed; if not, it installs Winget automatically.
2.	Check and Install **FFMPEG**
Tests if FFMPEG is installed; if missing, installs FFMPEG and sets the environment variable.
3.	Check and Install **Python 3.10**
Ensures Python 3.10 is installed; if not, installs it and sets the environment variable.

**Finally, it informs the user to reboot the system.**


**Batch Script 2** _2nd-part-install-git-ultrasinger-with-modules-musescore.bat_
[2nd-part-install-git-ultrasinger-with-modules-musescore.bat.TXT](https://github.com/user-attachments/files/17591385/2nd-part-install-git-ultrasinger-with-modules-musescore.bat.TXT) download and remove the ".TXT"

1.	Re-check and Install **Winget**
Confirms that Winget is truly installed; if not, installs it, or skips this step if already present.
2.	Check and Install **PIP**
Verifies if PIP is installed and installs it if necessary.
3.	Check and Install **GIT**
Ensures GIT is installed; if not, it installs GIT.
4.	Check and Install **UltraSinger**
Checks if UltraSinger is installed and installs it if needed.
5.	Check and Optional Install of **MuseScore**
Verifies if MuseScore is installed; if not, it asks the user whether to install MuseScore.
6.	Install Python Packages from requirements.txt
Installs all required packages listed in requirements.txt.
7.	Check and Install **CUDA**
Checks if a GPU is available. If yes, installs CUDA; otherwise, sets up **CPU** support.
8.	**Additional Python Packages**
Installs missing Python packages from additional-requirements.txt.

at the end it uninstalls torch and reinstalls torch - I do not know why it does not work without that step. 

File with python modules that seme to be missing
[additional-requirements.txt](https://github.com/user-attachments/files/17591392/additional-requirements.txt)

Batch workflow 
[batch_generate_karaoke_songs_from_youtube_playlist_with_ultrasinger.bat.TXT](https://github.com/user-attachments/files/17591395/batch_generate_karaoke_songs_from_youtube_playlist_with_ultrasinger.bat.TXT)
